### ISSUE TITLE
add a ratchet option to dataLim update functions and associated autoscale functions

### ISSUE THREAD

**Issue by ianhi** [matplotlib/matplotlib#18431]
### Problem

I think this was best stated by @mwaskom in https://github.com/matplotlib/matplotlib/issues/17331#issuecomment-624125670 

> What I really want from explicit limits / autoscaling is something closer to a ratchet: I want to set limits that won't hide to-be-added data that lie outside of the limits but will also not contract when to-be-added data lie inside the limits (minus the margin).

I am regularly running into this when making interactive plots. Trying to mix plot elements that may update with others that may not and some artists work with `ax.relim` while others do not. So it would be awesome if in addition to ratcheting behavior for `set_{x,y}lim` it was also implemented for the bbox dataLim updating methods.

### Proposed Solution

add `ratchet_x` and `ratchet_y` options to:
`ax.relim`
`ax.autoscale_view`
and the bbox updating methods:
https://github.com/matplotlib/matplotlib/blob/ab45bcdff003acd15d178a86a194e4002dca6a9f/lib/matplotlib/transforms.py#L885
https://github.com/matplotlib/matplotlib/blob/ab45bcdff003acd15d178a86a194e4002dca6a9f/lib/matplotlib/transforms.py#L849
https://github.com/matplotlib/matplotlib/blob/c9bf76ccec1581f42074a4fcb775144cbd058b02/src/_path.h#L368
https://github.com/matplotlib/matplotlib/blob/c9bf76ccec1581f42074a4fcb775144cbd058b02/src/_path.h#L349

I think the only place where the logic would actually need to be changed is in the `update_limits` function
and `ratchet` options to
`ax.set_xlim`
`ax.set_ylim`

### Additional context and prior art

A user can get this behavior in python by themselves like this:
```python
cur_ylims = ax.get_ylim()
ax.relim()
new_lims = [ax.dataLim.y0, ax.dataLim.y0+ax.dataLim.height]
ax.set_ylim(min(new_lims[0], cur_ylims[0]), max(new_lims[1], cur_ylims[1]))
ax.set_ylim(new_lims)
```
but this is likely less performant than if this was driven all the way down into the matplotlib c++ (I think?) and is not very convenient.

**Comment by jklymak** [comment]
I think you just want this at the high-level functions, not down into `_path.h`.  `update_path_extents` etc have to return the actual extents, not ratcheted versions. These functions do not know anything about the global limits

**Comment by ianhi** [comment]
feature request is still wanted

**Comment by jklymak** [comment]
I'm not following the use case here.  Can you give a set of calls that illustrate the workflow?

**Comment by mwaskom** [comment]
In the following snippet, the final x limits are tight around (2, 3), ignoring the limits I set previously:

```python
f, ax = plt.subplots()
ax.set_xlim(0, 4, auto=True)
ax.plot([2, 3], [1, 2])
```

But if I set `auto=False` and a subsequent call places data outside of those limits, I can't see it:

```python
f, ax = plt.subplots()
ax.set_xlim(0, 4, auto=False)
ax.plot([5, 6], [1, 2])
```

It would be helpful to have some way of saying "make the limits no less than (0, 4), but continue to auto-expand to additional data if necessary".

**Comment by jklymak** [comment]
I think I understood that.  What I don't understand is why you explicitly would want to set the xlim before adding data?

**Comment by mwaskom** [comment]
Because a third party library might want to create a plot, set axis limits that work with that plot, and then return the axis to a user who may add additional artists.

(The specific usecase for seaborn is that its "categorical" functions set axis limits from -0.5 to +0.5 the min and max category indices, because matplotlib's multiplicative margin approach doesn't work great with categorical data. But that's just one example of the general usecase).

**Comment by jklymak** [comment]
Thanks.  If the idea is to give third-party libraries flexibility, then that opens up the API to be a bit looser or cryptic than if we expected users to do this often. 

I wonder if we did something like `ax.set_xlim(0, 4, auto='lockmin')` and inside `set_xlim` we passed auto down into the auto-lim machinery as a two-tuple `auto=(0, 4)`.    The auto-lim machinery isn't nice, so I'm not sure how easy this would be to get to work robustly.   I prefer that to clashing keywords.

**Comment by jklymak** [comment]
BTW, what happens if I explicitly call `ax.set_xlim(20, 24)`?  Does it erase the previous window?  Expand the window to include the new limits?

**Comment by mwaskom** [comment]
Calling `set_xlim` again changes the xlimits to the new values. The relevant parameter here only controls *auto* scaling (really auto _limiting_ but that's the long-standing term).

**Comment by rcomer** [comment]
Does `update_datalim` do what you want?

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.update_datalim([(0, 0), (4, 0)], updatex=True, updatey=False)
ax.plot([2, 3], [1, 2])
```
![test](https://github.com/matplotlib/matplotlib/assets/10599679/122d15e6-19a9-49c5-af49-64d379fb716f)

```python
ax.plot([2, 6], [1, 2])
```
![test](https://github.com/matplotlib/matplotlib/assets/10599679/78e943b2-69eb-4720-82ee-0124e2b1862f)

**Comment by jklymak** [comment]
Wow, great find....  Obscure, but I think it is as described in the original request?

**Comment by mwaskom** [comment]
It's not exactly an answer to the request because if you want specific axis limits you need to work them out based on the axis margins, but yeah the basic mechanism seems to have the "ratchet" property that is desired.

**Comment by jklymak** [comment]
OK, `update_datalim` works by pretending that data was added with the given boundaries.  Conceptually that is a bit different than what is being asked for here, but achieves the same thing.

We could 1) add a ratchet to `set_xlim`.  2) add a "ignore_margins" to `update_datalim` (which would do the math to figure out what the data lim should be to get the requested data lim).  3) don't do anything.

**Comment by rcomer** [comment]
(2) doesn’t seem quite right to me, if I’ve understood.  I think you’re saying we set a shorter data range than requested so that the range plus the padding from the margins gives the requested axes limits.  But if we do that and a subsequent function moves one end of the data limits then that subsequent call would change the padding at the other end (because the range has increased).

**Comment by jklymak** [comment]
That's true   So this would still be a feature request if folks really didn't want the margins.
